sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java

```diff
@@ -70,21 +77,55 @@ public GeneratedColumns(InputFactory inputFactory,
         toValidate = validation == Validation.NONE ? Collections.emptyMap() : new HashMap<>();
         InputFactory.Context<CollectExpression<T, ?>> ctx = inputFactory.ctxForRefs(txnCtx, refResolver);
         toInject = new HashMap<>();
+
         for (Reference colWithDefault : allDefaultExpressionColumns) {
             if (!presentColumns.contains(colWithDefault)) {
                 toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
             }
         }
-        for (GeneratedReference generatedCol : allGeneratedColumns) {
-            if (presentColumns.contains(generatedCol)) {
-                if (validation == Validation.VALUE_MATCH) {
-                    toValidate.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
+
+        maybeInjectNested = new HashMap<>();
+        Set<String> topLevelPresentColumns = presentColumns
+            .stream()
+            .map(Reference::column)
+            .filter(ColumnIdent::isTopLevel)
+            .map(ColumnIdent::name)
+            .collect(Collectors.toSet());
+
+        allGeneratedColumns.forEach(
+            r -> processGeneratedExpressionColumn(
```

Comment:
Please keep in mind that the `GeneratedColumns` constructions shouldn't become too heavy as this is used for single inserts as well.. 

Lambdas that capture a closure tend to have a bit more bootstrapping overhead than a regular loop.